File: crates/db/src/entity/submission.rs
//! SeaORM `submission` entity — maps the `submission` table and its relations.

use async_trait::async_trait;
use sea_orm::{DeriveActiveEnum, EnumIter, Set, entity::prelude::*};
use serde::{Deserialize, Serialize};
use serde_repr::{Deserialize_repr, Serialize_repr};

#[sea_orm::model]
#[derive(Debug, Clone, PartialEq, Eq, DeriveEntityModel, Serialize, Deserialize)]
#[sea_orm(table_name = "submissions")]
pub struct Model {
    #[sea_orm(primary_key)]
    pub id: i64,
    pub content: String,
    pub status: Status,
    pub challenge_id: i64,
    pub user_id: i64,
    pub team_id: Option<i64>,
    pub game_id: Option<i64>,
    pub created_at: i64,

    #[sea_orm(default_value = 0)]
    pub pts: i64,
    #[sea_orm(default_value = 0)]
    pub rank: i64,
    #[sea_orm(belongs_to, from = "challenge_id", to = "id", on_delete = "Cascade")]
    pub challenge: BelongsTo<super::challenge::Entity>,
    #[sea_orm(belongs_to, from = "user_id", to = "id", on_delete = "Cascade")]
    pub user: BelongsTo<super::user::Entity>,
    #[sea_orm(belongs_to, from = "team_id", to = "id", on_delete = "Cascade")]
    pub team: BelongsTo<Option<super::team::Entity>>,
    #[sea_orm(belongs_to, from = "game_id", to = "id", on_delete = "Cascade")]
    pub game: BelongsTo<Option<super::game::Entity>>,
}

#[derive(
    Clone,
    Debug,
    Default,
    PartialEq,
    Eq,
    Serialize_repr,
    Deserialize_repr,
    EnumIter,
    DeriveActiveEnum,
    utoipa::ToSchema,
)]
#[sea_orm(rs_type = "i32", db_type = "Integer")]
#[repr(i32)]
pub enum Status {
    #[default]
    Pending   = 0,
    Correct   = 1,
    Incorrect = 2,
    Cheat     = 3,
    Expired   = 4,
    Duplicate = 5,
}

#[async_trait]
impl ActiveModelBehavior for ActiveModel {
    /// SeaORM lifecycle hook executed before insert/update.
    async fn before_save<C>(mut self, _db: &C, insert: bool) -> Result<Self, DbErr>
    where
        C: ConnectionTrait, {
        let ts = time::OffsetDateTime::now_utc().unix_timestamp();

        if insert {
            self.created_at = Set(ts);
        }

        Ok(self)
    }
}

File: web/src/models/submission.ts
export type SubmissionView = {
  id: number;
  content: string;
  status: Status;
  user_id: number;
  user_name: string;
  user_avatar_hash: string | null;
  challenge_id: number;
  challenge_title: string;
  challenge_category: number;
  team_id: number | null;
  team_name: string | null;
  team_avatar_hash: string | null;
  game_id: number | null;
  game_title: string | null;
  pts: number;
  rank: number;
  created_at: string;
  processing_at: string | null;
  checked_at: string | null;
};

export type SubmissionSummary = Omit<SubmissionView, "content">;

export enum Status {
  Queued = "queued",
  Processing = "processing",
  Correct = "correct",
  Incorrect = "incorrect",
  Cheat = "cheat",
  Expired = "expired",
  Duplicate = "duplicate",
}
